ipaserver/install/ipa_cert_fix.py

```diff
@@ -0,0 +1,244 @@
+
+
+def expired_certs(now):
+    return expired_dogtag_certs(now), expired_ipa_certs(now)
+
+
+def expired_dogtag_certs(now):
+    """
+    Determine which Dogtag certs are expired, or close to expiry.
+
+    Return a list of (cert_id, cert) pairs.
+
+    """
+    certs = []
+    db = NSSDatabase(nssdir=paths.PKI_TOMCAT_ALIAS_DIR)
+
+    for certid, nickname in [
+        ('sslserver', 'Server-Cert cert-pki-ca'),
+        ('subsystem', 'subsystemCert cert-pki-ca'),
+        ('ca_ocsp_signing', 'ocspSigningCert cert-pki-ca'),
+        ('ca_audit_signing', 'auditSigningCert cert-pki-ca'),
+    ]:
+        cert = db.get_cert(nickname)
+        if cert.not_valid_after <= now:
+            certs.append((certid, cert))
+
+    return certs
+
+
+def expired_ipa_certs(now):
+    """
+    Determine which IPA certs are expired, or close to expiry.
+
+    Return a list of (IPACertType, cert) pairs.
+
+    """
+    certs = []
+
+    # IPA RA
+    cert = x509.load_certificate_from_file(paths.RA_AGENT_PEM)
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.IPARA, cert))
+
+    # Apache HTTPD
+    db = NSSDatabase(nssdir=paths.HTTPD_ALIAS_DIR)
+    cert = db.get_cert('Server-Cert')
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.HTTPS, cert))
+
+    # LDAPS
+    ds_dbdir = dsinstance.config_dirname(
+        installutils.realm_to_serverid(api.env.realm))
+    db = NSSDatabase(nssdir=ds_dbdir)
+    cert = db.get_cert('Server-Cert')
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.LDAPS, cert))
+
+    # KDC
+    cert = x509.load_certificate_from_file(paths.KDC_CERT)
+    if cert.not_valid_after <= now:
+        certs.append((IPACertType.KDC, cert))
+
+    return certs
+
+
+def print_intentions(dogtag_certs, ipa_certs):
+    print("The following certificates will be renewed: ")
+    print()
+
+    for certid, cert in dogtag_certs:
+        print_cert_info("Dogtag", certid, cert)
+
+    for certtype, cert in ipa_certs:
+        print_cert_info("IPA", certtype.value, cert)
+
+
+def print_cert_info(context, desc, cert):
+    print("{} {} certificate:".format(context, desc))
+    print("  Subject: {}".format(DN(cert.subject)))
+    print("  Serial:  {}".format(cert.serial_number))
+    print("  Expires: {}".format(cert.not_valid_after))
+    print()
+
+
+def run_cert_fix(certs, extra_certs):
+    ldapi_path = (
+        paths.SLAPD_INSTANCE_SOCKET_TEMPLATE
+        % '-'.join(api.env.realm.split('.'))
+    )
+    cmd = [
+        'pki-server',
+        'cert-fix',
+        '--ldapi-socket', ldapi_path,
+        '--agent-uid', 'ipara',
+    ]
+    for certid, _cert in certs:
+        cmd.extend(['--cert', certid])
+    for _certtype, cert in extra_certs:
+        cmd.extend(['--extra-cert', str(cert.serial_number)])
+    subprocess.check_call(cmd)
+
+
+def replicate_dogtag_certs(certs):
+    for certid, _oldcert in certs:
+        cert_path = "/etc/pki/pki-tomcat/certs/{}.crt".format(certid)
```

Comment:
Yes, it is.  This is not documented, because the PKI cert-fix tool did not consider that the caller would need to do anything further with those certs.  I will (some day soon) update the cert-fix documentation to make this part of the contract, as well as documenting my other enhancements.

Reply:
ok, this approach works for me.